### PR TITLE
Shed traffic when server receives SIGTERM

### PR DESCRIPTION
**Pull request by ferozco**
## Issue
When the server received a SIGTERM (typically from k8s when rolling the pod) we would immediately shutdown the underlying undertow server, effectively terminating any inflight requests. This was problematic because frequently re-deploying a service would cause errors for upstream callers

## Summary
Shed traffic when server receives SIGTERM

Instead of shutting down the server right away, we start sending back 503s effectively re-directing traffic to other pods. The server will remain in this state until k8s sends a SIGKILL after `terminationGracePeriodSeconds` at which point the pod will be completely killed

### CODE REVIEW

**Comment by markelliot** [comment]
I think we need/want to use Undertow's https://undertow.io/javadoc/2.1.x/io/undertow/server/handlers/GracefulShutdownHandler.html rather than this custom built option, and that upon shutdown (using https://undertow.io/javadoc/2.1.x/io/undertow/server/handlers/GracefulShutdownHandler.ShutdownListener.html) we should stop the Undertow server.

As is, the PR breaks the semantic that Barista will come to a stop after receiving a SIGTERM.

**Comment by markelliot** [comment]
I can't find great docs for this, so to save you some searching: I think we'd roughly:
 - register and stash away for later a GracefulShutdownHandler
 - register a Runtime shutdown hook that tells the GracefulShutdownHandler to shutdown + after invoking GSH#shutdown using awaitShutdown(long) to wait a maximum amount of time (say 2 minutes) for requests to complete and then still calling Undertow#shutdown
 - register a listener on the GSH's shutdown state that shuts down Undertow

**Comment by ferozco** [comment]
@markelliot mind taking another look?

File: barista/src/main/java/com/markelliot/barista/Server.java [hunk 1]
```diff
@@ -62,7 +68,15 @@ private void start() {
      * included automatically.
      */
     public void stop() {
-        undertow.stop();
+        shutdownHandler.shutdown();
+        shutdownHandler.addShutdownListener(shutdownSuccessful -> undertow.stop());
```
Comment (markelliot) [inline barista/src/main/java/com/markelliot/barista/Server.java:72]: I think we ought to do this before calling shutdown to be sure that the listener is definitely called

Comment (ferozco) [reply barista/src/main/java/com/markelliot/barista/Server.java:72]: you can only add shutdown hooks after you've told the handler to shutdown.

https://github.com/undertow-io/undertow/blob/master/core/src/main/java/io/undertow/server/handlers/GracefulShutdownHandler.java#L169-L171

**Review by markelliot — COMMENTED** [review]
I think one minor thing

File: barista/src/main/java/com/markelliot/barista/Server.java (hunk 1 above)
Comment (markelliot) [reply barista/src/main/java/com/markelliot/barista/Server.java:72]: Oh that’s weird

**Review by markelliot — APPROVED** [review]

**Pull request merged**